Group function

minad opened the pull request:
Corresponding PR for #458. I implemented the `group-function`. You may want to pull the first four simplification commits directly. In order to test the grouping feature, you can `eval-buffer` consult.el from https://github.com/minad/consult/pull/234.

consult-mode-command:
![cmd](https://user-images.githubusercontent.com/50754/109176443-1f0a6f80-7787-11eb-92f8-326d615f04af.png)

minad commented:
@clemera If you want to test this, you can `eval-buffer` consult.el from https://github.com/minad/consult/commit/3a0613109635f5edcbeb4299059f95a21e0aa590.

clemera commented:
Thank you, I did not expect that you want to make it customizable hence my suggestion to add it as metadata. I thought of it as an affixation/annotation which can't be configured, too. I will look at this tomorrow, is this done from your side?

minad commented:
I added the `selectrum-group-candidates-function` since this way it is nicely aligned with the `selectrum-preprocess-candidates-function`, see for example `selectrum--preprocess`. Furthermore the grouping functions is needed twice: 1. to group the candidates during preprocessing, 2. when adding the headlines to the list of displayed candidate strings. If you prefer, we can also rename it to `selectrum--group-candidates-function` in order to not commit too early on a public API.

I am done with this PR. From my side it is ready to merge. I already implemented full support for this feature in Consult too. If you find things that should be corrected, please let me know.

minad commented:
You are indeed right that it is better to remove the group function as a public variable. It should be treated the same way as the annotation function.

clemera commented:
Thanks, I don't have much time right now, but it looks you have also done some nice refactorings which is great. I'm not sure about the removing the callback comment but it sounds like you have a plan there ;)

minad commented:
> Thanks, I don't have much time right now, but it looks you have also done some nice refactorings which is great.

Yes, I usually try to do a few preparatory refactorings before fitting in a new feature. This way stuff stays maintainable :)

> I'm not sure about the removing the callback comment but it sounds like you have a plan there ;)

I also added a few notes here and there. Most of them are already gone (and not visible here anymore) since I made the refactorings, either because the refactorings were necessary for the new feature or made it easier to integrate.

I didn't touch this callback dance with the candidate insertion for now. Nevertheless I would love to see this part of the code simplified. Maybe I will do this in another PR - in any case it will not be part of this PR. I don't want to throw too many things together.

clemera commented:
Okay good, I'm happy to have another PR for that if you like, I'm not sure I would like to remove it but maybe I will change my mind when I see how you would like to do it.

minad commented:
See also https://github.com/oantolin/icomplete-vertical/pull/25

clemera commented on selectrum.el:
```diff
@@ -1129,6 +1151,23 @@ defaults to the current one and MAX which defaults to
+  "Preprocess CANDIDATES list.
+The preprocessing applies the `selectrum-preprocess-candidates-function'
+and the `x-group-function'."
+  (setq selectrum--preprocessed-candidates
```
Can be changed to `setq-local`

minad replied:
Fixed now. I missed this comment before.

clemera commented:
Looks all good to me, I also like the refactorings you have done, again thanks for that! Here are my observations:

- The ido config from the wiki uses `selectrum--get-meta` and needs to be updated
- There can be candidates off the screen when jumping to the last candidate (similar to the problem we have with `C-x 8 RET`)
- I'm not sure about the `selectrum-group-format` + faces combo. The format has to use the faces so maybe we shouldn't allow to not using them, maybe the faces could be included as anonymous faces. Another point is that I would prefer `:inherit` for the default values (from shadow?), the separators were barely visible with my current theme for example and the height is to small for me.

Regarding the comment to remove the insertion function: For now I'm assuming we still may allow arbitrary insertion functions and the function needs to tell us some information for formatting the candidates correctly. I'm open to change this approach in the future or let you convince me that we can do it in better ways but before I don't have a clearer picture I wouldn't like to have the comment there.

minad commented:
> The ido config from the wiki uses selectrum--get-meta and needs to be updated

I decided to reintroduce it for the reasons mentioned in the commit.

> There can be candidates off the screen when jumping to the last candidate (similar to the problem we have with C-x 8 RET)

Yes, this is an unfortunate issue. It is very much related to the window resizing problems. I would like this to be fixed in a separate PR. Is that okay for you?

> I'm not sure about the selectrum-group-format + faces combo. The format has to use the faces so maybe we shouldn't allow to not using them, maybe the faces could be included as anonymous faces inside the format variable.

I would very much like to keep it as is. It is necessary to have separate faces since this allows themes to adjust the definitions. If we use inline face definitions, this is not possible.

> Another point is that I would prefer :inherit for the default values (from shadow?), the separators were barely visible with my current theme for example and the height is to small for me.

Okay, we can do that. I like these small separator lines, but maybe this is too opinionated.

> I wouldn't like to have the comment there.

I will remove the note commit and reorganize the commits slightly.

clemera commented:
> Yes, this is an unfortunate issue. It is very much related to the window resizing problems. I would like this to be fixed in a separate PR. Is that okay for you?

Absolutely, we have to think about how to handle resizing for these cases.

> I would very much like to keep it as is. It is necessary to have separate faces since this allows themes to adjust the definitions. If we use inline face definitions, this is not possible.

Makes sense, it is better to keep it that way.

minad commented:
I made these changes:

* Reorganized the commits (please do not stash merge, the history is useful)
* Updated the faces, inheriting from shadow now
* Removed note comments

minad commented:
Okay, now it should really be ready.

clemera commented:
Thank you, usually I prefer to squash but if the commits are neatly organized like this I won't object to merge them as is :)

minad commented:
Thanks! 🎉

clemera commented:
Hm, somehow the CI detection I usually rely on wasn't working in this case, fixed the formatting in d1447e9a5e0135b5a7fff1600ea0577e5ef45742

minad commented:
I also relied on the CI. Why is it broken?

clemera commented:
It did show the red x on the commit after the merge and when testing locally via `make lint` I saw the issue were the long lines.

clemera commented:
Oh, it is still broken, probably because support for older Emacsen I will take a look.

minad commented:
I have package-lint enabled always. Usually it warns me of problems. 

But if there is a problem - can we please just drop support for 25? There is no point. You can get 26 everywhere and now even Debian stable has 27. I think the selectrum readme mentions other problems with 25.

clemera commented:
Okay we can drop support for it but I haven't done the setup for CI and don't know how to do it, probably the Dockerfile has to be adjusted.

clemera commented:
BTW the problem is `mapcan` which wasn't available before Emacs 26.

minad commented:
Okay, I see. I wonder why package-lint did not warn. Mapcan is more efficient. I would like to keep this. Note that Consult requires emacs 26 and it is the only producer of x-group-functions by now :) But seriously, just drop 25.
